src/wp-includes/functions.php

```diff
@@ -2610,6 +2687,31 @@ function wp_unique_filename( $dir, $filename, $unique_filename_callback = null )
 	return apply_filters( 'wp_unique_filename', $filename, $ext, $dir, $unique_filename_callback );
 }
 
+/**
+ * Helper function to test if each of an array of file names could conflict with existing files.
+ *
+ * @since 5.8.1
+ * @access private
+ *
+ * @param string[] $filenames Array of file names to check.
+ * @param string   $dir       The directory containing the files.
+ * @param array    $files     An array of existing files in the directory. May be empty.
+ * @return bool True if the tested file name could match an existing file, false otherwise.
+ */
+function _wp_check_alternate_file_names( $filenames, $dir, $files ) {
```

Comment:
Hehe, not primary but yes, avoid if possible. The primary reason is to not run `wp_unique_filename()` recursively.

Reply:
> The primary reason is to not run `wp_unique_filename()` recursively.

This is where I very much disagree.

`wp_unique_filename()` *should* be called recursively to ensure the best method for checking filenames is always used, and to ensure that its filter is fired for each proposed filename, including the alternate format filenames that otherwise only `wp_unique_filename()` itself would check.

My next path has a test case that explains this better.

Reply:
> should be called recursively...

This introduces a bug/edge case, see https://core.trac.wordpress.org/ticket/53668#comment:25. All alternate filenames should be checked together at the same time or "gaps" in the numbering (i.e. file-1, file-2, file-4) will result in overwriting.

Reply:
> This introduces a bug/edge case, see https://core.trac.wordpress.org/ticket/53668#comment:25. All alternate filenames should be checked together at the same time or "gaps" in the numbering (i.e. file-1, file-2, file-4) will result in overwriting.

I could be wrong, but I don't believe my code has this bug as it re-checks the original and alternate extensions set should a new version be proposed on exit from the protected loop.

Do you have a test case for this that breaks my latest patch?